quinn-proto/src/stream.rs

```diff
@@ -400,97 +399,98 @@ pub enum RecvState {
-            written_offset: 0,
+            data: BinaryHeap::new(),
         }
     }
 
     /// Whether `read` will return nonzero
     pub fn blocked(&self) -> bool {
-        let mask = !0 >> self.written_offset;
-        self.written.front().map_or(true, |x| x & mask == mask)
+        match self.data.peek() {
+            Some(first) => first.offset > self.offset,
+            None => true,
+        }
     }
 
-    /// Leading written bytes
-    fn prefix_len(&self) -> usize {
-        for i in 0..self.written.len() {
-            let x = self.written[i];
-            if x == 0 || (i == 0 && x << self.written_offset == 0) {
-                continue;
+    pub fn read(&mut self, buf: &mut [u8]) -> usize {
+        let mut read = 0;
+        loop {
+            if self.consume(buf, &mut read) {
+                self.data.pop();
+            } else {
+                break;
+            }
+            if read == buf.len() {
+                break;
             }
-            return (i * 8 + x.leading_zeros() as usize) - self.written_offset as usize;
         }
-        self.written.len()
+        read
     }
 
-    pub fn read(&mut self, buf: &mut [u8]) -> usize {
-        let (a, b) = self.data.as_slices();
-        let available = self.prefix_len();
-        let a_len = a.len().min(available);
-        let (a, b) = (&a[0..a_len], &b[0..(available - a_len)]);
-        let a_n = a.len().min(buf.len());
-        buf[0..a_n].copy_from_slice(&a[0..a_n]);
-        let b_n = b.len().min(buf.len().saturating_sub(a.len()));
-        buf[a_n..(a_n + b_n)].copy_from_slice(&b[0..b_n]);
-        let n = a_n + b_n;
-
-        self.offset += n as u64;
-        self.data.drain(0..n);
-        let q = n / 8;
-        let r = n % 8;
-        let carry = (self.written_offset as usize + r) / 8;
-        self.written.drain(0..(q + carry));
-        self.written_offset = (self.written_offset as usize + r - carry * 8) as u8;
-
-        n
+    // Read as much from the first chunk in the heap as fits in the buffer.
+    // Takes the buffer to read into and the amount of bytes that has already
+    // been read into it. Returns whether the first chunk has been fully consumed.
+    fn consume(&mut self, buf: &mut [u8], read: &mut usize) -> bool {
+        let mut chunk = if let Some(chunk) = self.data.peek_mut() {
+            chunk
+        } else {
+            return false;
+        };
+
+        // If this chunk is either after the current offset or fully before it,
+        // return directly, indicating whether the chunk can be discarded.
+        if chunk.offset > self.offset {
+            return false;
+        } else if (chunk.offset + chunk.bytes.len() as u64) < self.offset {
+            return true;
+        }
+
+        // Determine `start` and `len` of slice to read from chunk
+        let start = if chunk.offset < self.offset {
+            (self.offset - chunk.offset) as usize
+        } else {
+            0
+        };
```

Comment:
We already returned if `chunk.offset > self.offset`, so this branch isn't needed.

Reply:
Yes, this is still needed. This is the inverse condition, where the next chunk starts before the current assembler offset (but the chunk extends beyond the current offset). For example, the `assemble_overlapping` test case.

Reply:
I understand, but the branch isn't necessary. The test case still passes with
```rust
let start = (self.offset - chunk.offset) as usize;
```
because at this point in control flow it is guaranteed that chunk.offset <= self.offset, so there will never be underflow.

Reply:
Ah, I misunderstood your comment. Fixed!